### PR TITLE
fix: #1015 Notification issues when the text is too long

### PR DESCRIPTION
**Pull request by J0taFerreira**
- added breakpoint for the ids shown
- limit the max of visible show ids to 5

### CODE REVIEW

File: frontend/src/store/screenSizeState.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { sideNavState } from "@/store/SideNavState";
+import { setNotificationMsg, newMsgList, newFormData, newSeverity } from "./NotificationState";
+
+export let screenSize = window.innerWidth;
```
Comment (MCatherine1994) [inline frontend/src/store/screenSizeState.ts:4]: Hi Jota, if you export this variable and share it with other components, it has to be a reactive variable, not a local changeable variable.  Cause whenever you change its value you want all other components receive the new value and re-render the page.

File: frontend/src/store/NotificationState.ts [hunk 2]
```diff
@@ -3,27 +3,78 @@
...
+import { isDesktop, screenSize } from "@/store/screenSizeState";
+import type { Severity } from "@/enum/SeverityEnum";
 
 const defaultNotification = {
     success: "",
     warn: "",
     error: "",
 };
 
+export const newMsgList = ref();
```
Comment (MCatherine1994) [inline frontend/src/store/NotificationState.ts:17]: and we can't use one reactive variable to support multiple notification message, in the case we have both successful and warning message, the auto adjustment on the notification message will fail
Screen Shot 2023-11-14 at 11 04 54 AM

Comment (J0taFerreira) [reply frontend/src/store/NotificationState.ts:17]: Hello! I'll fix it. Thanks!

File: frontend/src/store/screenSizeState.ts (hunk 1 above)
Comment (J0taFerreira) [reply frontend/src/store/screenSizeState.ts:4]: Sure!

File: frontend/src/store/NotificationState.ts [hunk 3]
```diff
@@ -3,27 +3,78 @@
     It is structured this way so it doesn't stack multiple messages with the same severity (by design).
     It is intended to be used in conjunction with the NotificationStack component.
 */
-import type { Severity } from "@/enum/SeverityEnum";
+
 import { ref } from "vue"
+import { isDesktop, screenSize } from "@/store/screenSizeState";
+import type { Severity } from "@/enum/SeverityEnum";
 
 const defaultNotification = {
```
Comment (MCatherine1994) [inline frontend/src/store/NotificationState.ts:11]: So instead of creating t here new variables (newMsgList, newFormData, newSeverity), we can update the structure of our notification variable.

```
const defaultNotification = {
    success: {msg: "", forestClientNumberList: ""},
    warn: {msg: "", forestClientNumberList: ""},
    error: {msg: "", forestClientNumberList: ""},
}
```

And when you push notification, we store both message and the forest client number list:
```
export const pushNotification = (severity: Severity, message: string, forestClientNumberList: Array<string>) => {
     notifications.value[severity].msg = message;
     notifications.value[severity].forestClientNumberList = forestClientNumberList
 }
```

File: frontend/src/store/NotificationState.ts [hunk 4]
```diff
@@ -3,27 +3,78 @@
...
+    const hiddenIdList = msgList.splice(idPerMsg());
+
+    const isPlural = msgList.length === 1 && hiddenIdList.length === 0 ? 'ID' : 'IDs'
+    const msgByType = {
+        success: `was successfully added with Client ${isPlural}:`,
+        warn: `already exists with Client ${isPlural}:`,
+        error: `was not added with Client ${isPlural}:`
+    };
+
+    const setNotificationMsg = `
```
Comment (MCatherine1994) [inline frontend/src/store/NotificationState.ts:72]: we better not name a variable the same as a method, please rename it

File: frontend/src/store/NotificationState.ts (hunk 2 above)
Comment (MCatherine1994) [reply frontend/src/store/NotificationState.ts:17]: we can remove these three new variables

File: frontend/src/store/NotificationState.ts [hunk 5]
```diff
@@ -3,27 +3,78 @@
...
+    } else if(screenSize <= 1290 && isDesktop() || screenSize <= 960) {
+        return 4;
+
+    } else {
+        return 5;
+
+    }
+};
+
+export const setNotificationMsg = (msgList: string[], formData: any, severity: Severity) => {
```
Comment (MCatherine1994) [inline frontend/src/store/NotificationState.ts:57]: we don't use formData parameter at all, and please rename the msgList, I feel it's actually the forest client number list?

File: frontend/src/store/screenSizeState.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+window.addEventListener("resize", (event) => {
+    screenSize = window.innerWidth;
+
+    //update the message when the screensize change
+    if(newMsgList.value) {
+        setNotificationMsg(newMsgList.value, newFormData.value, newSeverity.value);
```
Comment (MCatherine1994) [inline frontend/src/store/screenSizeState.ts:11]: and here, you can create a for loop:

Object.keys(notifications.value).forEach((severity) => {
    if(notifications.value[severity].msg &&  notifications.value[severity].forestClientNumberList > 1) {
        setNotificationMsg(notifications.value[severity].forestClientNumberList, severity);
    }
})

File: frontend/src/components/grantaccess/GrantAccess.vue [hunk 7]
```diff
@@ -28,7 +28,8 @@ import {
 
 import { IconSize } from '@/enum/IconEnum';
 import { Severity } from '@/enum/SeverityEnum';
-import { pushNotification } from '@/store/NotificationState';
+import { pushNotification, setNotificationMsg } from '@/store/NotificationState';
```
Comment (MCatherine1994) [inline frontend/src/components/grantaccess/GrantAccess.vue:31]: if we don't call pushNotification in this file, let's remove the import

File: frontend/src/store/NotificationState.ts [hunk 8]
```diff
@@ -3,27 +3,78 @@
...
+    const msgByType = {
+        success: `was successfully added with Client ${isPlural}:`,
+        warn: `already exists with Client ${isPlural}:`,
+        error: `was not added with Client ${isPlural}:`
+    };
+
+    const setNotificationMsg = `
+        ${formData} ${msgByType[severity]} ${fullIdList.join(', ')}
+        ${isPlural === 'IDs' ?
+        'and ' + hiddenIdList.length + ' more...'
```
Comment (MCatherine1994) [inline frontend/src/store/NotificationState.ts:75]: when the hiddenIdList is 0, there is no need to show "and 0 more..."

File: frontend/src/store/screenSizeState.ts (hunk 1 above)
Comment (ianliuwk1019) [reply frontend/src/store/screenSizeState.ts:4]: Use getter if you need screenSize value.

File: frontend/src/store/NotificationState.ts [hunk 9]
```diff
@@ -3,27 +3,78 @@
...
+        return 3;
+
+    } else if(screenSize <= 1290 && isDesktop() || screenSize <= 960) {
+        return 4;
+
+    } else {
+        return 5;
+
+    }
+};
```
Comment (ianliuwk1019) [inline frontend/src/store/NotificationState.ts:55]: The implementation requirement isn't clear here.
Did you discuss with business (or @OlgaLiber2 / @MCatherine1994 ) in the ticket that: hiding id(s) for the user is **acceptable**? 
Is it one of the acceptance criteria in the ticket? Please add this if its been discussed.
(We don't just make code works).
We need to ask, for example:
* how many of the ids can the field take, is there such requirement?
* if there is no limit, can we fit the ids in the rendering container conditionally (css style change)?
* if there is no limit, for mobile device, what else can we do to handle it better before we really chop the ids.
Please discuss these kind of questions before implementing it or before opening a PR if you already see a problem.

**Review by ianliuwk1019 — CHANGES_REQUESTED** [review]
Hi @Jota,
When implementing some logic, it is better to know if this logic should be in common place or it should be in specific place.
We don't want to mix common file/function/logic with business specific. It will confuse developers and resulting into hard to maintain code. Time consuming to refactor. Sometimes it is not easy but please keep that in mind.
For example in this case, the purpose of `NotificationState` is rather confusing me. I am not getting a sense it is for general purpose, but rather incorporate specific business concept embedded in there already (like using **formData**, **forestClientId**).
Also rather having `screenSizeState` state to change some notification message, it may be better to let screenSizeState alone and notify (since `ref` is reactive) whichever state or component needs to change to do its jobs.

File: frontend/src/store/NotificationState.ts (hunk 3 above)
Comment (J0taFerreira) [reply frontend/src/store/NotificationState.ts:11]: I'll use your suggestion

File: frontend/src/store/NotificationState.ts (hunk 5 above)
Comment (J0taFerreira) [reply frontend/src/store/NotificationState.ts:57]: I'll teste if it works without the formData

Comment (J0taFerreira) [reply frontend/src/store/NotificationState.ts:57]: the formData is used to get the user id but I'll double check

File: frontend/src/components/grantaccess/GrantAccess.vue (hunk 7 above)
Comment (J0taFerreira) [reply frontend/src/components/grantaccess/GrantAccess.vue:31]: sure, I'll remove it

File: frontend/src/components/grantaccess/GrantAccess.vue [hunk 10]
```diff
@@ -257,30 +259,13 @@ async function handleSubmit() {
...
         if (successList.length > 0) {
-            pushNotification(
-                Severity.success,
-                `${
-                    formData.value.userId
-                } was successfully added with Client IDs: ${successList.join(
-                    ', '
-                )}`
-            );
+            setNotificationMsg(successList, formData.value.userId, Severity.success);
```
[the remark this answers is not in the document]
Comment (J0taFerreira) [reply frontend/src/components/grantaccess/GrantAccess.vue:262]: it takes the userId, I'll refactor to be more clear

File: frontend/src/store/NotificationState.ts [hunk 11]
```diff
@@ -3,27 +3,78 @@
...
+
 import { ref } from "vue"
+import { isDesktop, screenSize } from "@/store/screenSizeState";
+import type { Severity } from "@/enum/SeverityEnum";
 
 const defaultNotification = {
     success: "",
     warn: "",
     error: "",
 };
```
[the remark this answers is not in the document]
Comment (J0taFerreira) [reply frontend/src/store/NotificationState.ts:15]: it handle a single message

File: frontend/src/store/NotificationState.ts (hunk 9 above)
Comment (J0taFerreira) [reply frontend/src/store/NotificationState.ts:55]: hiding the ids that don't fit is on the design

File: frontend/src/store/NotificationState.ts (hunk 5 above)
Comment (MCatherine1994) [reply frontend/src/store/NotificationState.ts:57]: user id you mean the forest client number? but the msgList are for the forest client number as well?

File: frontend/src/store/NotificationState.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (MCatherine1994) [reply frontend/src/store/NotificationState.ts:15]: yeah it handles single message for now, I think in the future if we need to handle multiple success message we can rethink about it, but we can focus on the the issue in the ticket for now.

File: frontend/src/store/NotificationState.ts (hunk 5 above)
Comment (J0taFerreira) [reply frontend/src/store/NotificationState.ts:57]: the formData is the user name and the msgList it the list with the client ids

Comment (MCatherine1994) [reply frontend/src/store/NotificationState.ts:57]: then let's just pass the user name, and name the parameter that we can understand what it's for

File: frontend/src/store/NotificationState.ts [hunk 12]
```diff
@@ -3,15 +3,27 @@
...
 import type { Severity } from "@/enum/SeverityEnum";
-import { ref } from "vue"
 
 const defaultNotification = {
     success: "",
     warn: "",
     error: "",
 };
 
+export const seeAll = reactive({
```
Comment (MCatherine1994) [inline frontend/src/store/NotificationState.ts:15]: Hi Jota, I got what you're trying to do here and I think it's a good idea to store the full message. But I wonder if we could combine the two variables "notifications" and "seeAll" together? like following? otherwise I feel it's a little bit hard to follow and understand, we should always try to keep the logic simple and reusable. Please let me know if you have any questions, thanks!!
```
const defaultNotifications = {
    success: {msg: "", fullMsg: "", showFullMsg: false},
    warn:  {msg: "", fullMsg: "", showFullMsg: false},
    error:  {msg: "", fullMsg: "", showFullMsg: false},
};

export const notifications = ref(
    JSON.parse(JSON.stringify(defaultNotification))
);

export const showFullNotificationMsg = (severity: Severity) => {
    notifications.value[severity].msg = notifications.value[severity].fullMsg
    notifications.value[severity].showFullMsg = true
}
```

File: frontend/src/components/grantaccess/GrantAccess.vue [hunk 13]
```diff
@@ -242,44 +243,13 @@ const composeAndPushNotificationMessages = (
...
-                warningIdList[0] === ''
-                    ? `already exists with the role ${
-                          getSelectedRole()?.role_name
-                      }`
-                    : `already exists with Client IDs: ${warningIdList.join(
-                          ', '
-                      )}`
-            }`
-        );
+        setNotificationMsg(warningIdList, username, Severity.warning );
```
Comment (MCatherine1994) [inline frontend/src/components/grantaccess/GrantAccess.vue:249]: let's remove the space between "warning" and ")"

**Comment by MCatherine1994** [comment]
the new style make the notification box really wide, is it just me? or this is expected maybe?
Screen Shot 2023-11-22 at 12 42 49 PM

File: frontend/src/store/NotificationState.ts (hunk 12 above)
Comment (J0taFerreira) [reply frontend/src/store/NotificationState.ts:15]: Hello, Catherine! I can do that

File: frontend/src/components/grantaccess/GrantAccess.vue (hunk 13 above)
Comment (J0taFerreira) [reply frontend/src/components/grantaccess/GrantAccess.vue:249]: sure, will do!

**Review by MCatherine1994 — APPROVED** [review]
